src/main.cpp

```diff
@@ -407,324 +264,129 @@ int main(int argc, const char* argv[]) {
-        AstVisitor().visit_program(*ast);
-
-        /// Check some rules that ast should follow
-        {
-            logger->info("Running semantic analysis visitor");
-            if (SemanticAnalysisVisitor().check(*ast)) {
-                return 1;
-            }
-        }
-
-        /// construct symbol table
-        {
-            logger->info("Running symtab visitor");
-            SymtabVisitor(update_symtab).visit_program(*ast);
-        }
-
-        /// use cnexp instead of after_cvode solve method
-        {
-            logger->info("Running CVode to cnexp visitor");
-            AfterCVodeToCnexpVisitor().visit_program(*ast);
-            ast_to_nmodl(*ast, filepath("after_cvode_to_cnexp", "mod"));
-        }
-
-        /// Rename variables that match ISPC compiler double constants
-        if (ispc_backend) {
-            logger->info("Running ISPC variables rename visitor");
-            IspcRenameVisitor(ast).visit_program(*ast);
-            SymtabVisitor(update_symtab).visit_program(*ast);
-            ast_to_nmodl(*ast, filepath("ispc_double_rename", "mod"));
-        }
-
-        /// GLOBAL to RANGE rename visitor
-        if (nmodl_global_to_range) {
-            // make sure to run perf visitor because code generator
-            // looks for read/write counts const/non-const declaration
-            PerfVisitor().visit_program(*ast);
-            // make sure to run the GlobalToRange visitor after all the
-            // reinitializations of Symtab
-            logger->info("Running GlobalToRange visitor");
-            GlobalToRangeVisitor(*ast).visit_program(*ast);
-            SymtabVisitor(update_symtab).visit_program(*ast);
-            ast_to_nmodl(*ast, filepath("global_to_range", "mod"));
-        }
-
-        /// LOCAL to ASSIGNED visitor
-        if (nmodl_local_to_range) {
-            logger->info("Running LOCAL to ASSIGNED visitor");
-            PerfVisitor().visit_program(*ast);
-            LocalToAssignedVisitor().visit_program(*ast);
-            SymtabVisitor(update_symtab).visit_program(*ast);
-            ast_to_nmodl(*ast, filepath("local_to_assigned", "mod"));
-        }
-
-        {
-            // Compatibility Checking
-            logger->info("Running code compatibility checker");
-            // run perfvisitor to update read/write counts
-            PerfVisitor().visit_program(*ast);
-
-            // If we want to just check compatibility we return the result
-            if (only_check_compatibility) {
-                return CodegenCompatibilityVisitor().find_unhandled_ast_nodes(*ast);
-            }
-
-            // If there is an incompatible construct and code generation is not forced exit NMODL
-            if (CodegenCompatibilityVisitor().find_unhandled_ast_nodes(*ast) && !force_codegen) {
-                return 1;
-            }
-        }
+        auto cg_driver = CodegenDriver(cfg);
+        auto success = cg_driver.prepare_mod(*ast);
 
         if (show_symtab) {
```

Comment:
Moving this after all the passes means that the symbol table printed would be the one updated from all the passes which is something not bad but different than before. I think it's still fine for me because most of the time we want to see all the variables in the symtab for debugging purposes. I don't know if @pramodk has a different opinion on this